configbp/config.go

```diff
@@ -0,0 +1,69 @@
+// Package configbp parses configurations per the baseplate spec.
+package configbp
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/reddit/baseplate.go/log"
+)
+
+// ExpandConfig performs baseplate-conformant configuration expansion.
+//
+// Currently, this includes expanding $FOO and ${FOO} with their respective environment variables.
+func ExpandConfig(rawConfig string) string {
+	expanded := os.ExpandEnv(rawConfig)
+	if expanded != rawConfig {
+		log.Debugf("Expanded configuration:\n%s", expanded)
+	}
+	return expanded
+}
+
+// ParseFile parses configuration from the file at the given path.
+//
+// Environment variables (e.g. $FOO and ${FOO}) are substituted from the environment before parsing.
+// The configuration is parsed into the target, which will typically be a pointer to a struct.
+//
+// Decoding is performed based on file extension:
+//   - YAML (i.e. ".yaml") files are decoded according to ParseYAML.
+func ParseFile(path string, ptr interface{}) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err // contains filename
+	}
+	defer f.Close() // safe to blindly close read-only files
+
+	switch ext := filepath.Ext(path); ext {
+	case ".yaml":
+		return ParseYAML(f, ptr)
+	default:
+		return fmt.Errorf("unsupported config extension %q", ext)
+	}
+}
+
+// ParseYAML parses YAML read from the given Reader with strict decoding enabled.
+//
+// Environment variables (e.g. $FOO and ${FOO}) are substituted from the environment before parsing.
+// The configuration is parsed into the target, which will typically be a pointer to a struct.
+func ParseYAML(reader io.Reader, ptr interface{}) error {
+	rawData := new(strings.Builder)
+	if _, err := io.Copy(rawData, reader); err != nil {
```

Comment:
at the very least, please use `io.LimitReader` with a sane limit to wrap `reader`.

I think it's doable to fork `os.Expand` to wrap an `io.Reader` instead of converting strings, but that's probably too hairy and not really worth doing.

Reply:
This kind of thing has led to things failing in production before, as things cross arbitrary thresholds.  I'd rather not.

Reply:
there are a lot of ways to handle that:

1. Define it as an exported global variable instead of a const, so if a service know that they are using a huge config file, they can change the value of the limit in their main function, before calling `baseplate.New`.
2. Implement `io.LimitReader` by ourselves, using `bufio.Reader`, and return a different error than eof when we hit the limit and there are still more data in the original reader. This way if the yaml parser really reads beyond the limit it will fail, and that failure will be output to the log.
3. The line-buffer idea you talked about is also worth considering, but I don't believe there's a limit to the line either, so it might still oom when reading from for example `/dev/zero`.

So you probably need 1+2 or 3, but really, don't just copy the whole file into memory.

Reply:
2: https://github.com/reddit/baseplate.go/pull/388

Reply:
I have introduced a line-based environment substitutor.